loopy/tools.py

```diff
@@ -862,4 +865,50 @@ def t_unit_to_python(t_unit, var_name="t_unit",
     else:
         return python_code
 
+
+def memoize_on_disk(func, key_builder_t=LoopyKeyBuilder):
+    from loopy.version import DATA_MODEL_VERSION
+    from functools import wraps
+    from pytools.persistent_dict import WriteOncePersistentDict
+    from loopy.translation_unit import TranslationUnit
+    from loopy.kernel import LoopKernel
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        from loopy import CACHING_ENABLED
+
+        if (not CACHING_ENABLED
+                or kwargs.pop("_no_memoize_on_disk", False)):
+            return func(*args, **kwargs)
+
+        transform_cache = WriteOncePersistentDict(
+            ("loopy-memoize-cache-"
+             f"{key_builder_t.__qualname__}-{key_builder_t.__name__}"
+             f"-v0-{DATA_MODEL_VERSION}"),
+            key_builder=key_builder_t())
+        cache_key = (args, kwargs)
+
+        try:
+            result = transform_cache[cache_key]
+            logger.debug(f"Function {func.__name__} returned from"
+                         " memoized result on disk.")
+            return result
+        except KeyError:
+            logger.debug(f"Function {func.__name__} not present"
+                         " on disk.")
+            if isinstance(args[0], LoopKernel):
+                proc_log_str = f"{func.__name__} on '{args[0].name}'"
+            elif isinstance(args[0], TranslationUnit):
+                proc_log_str = f"{func.__name__} on '{args[0].entrypoints}'"
```

Comment:
This now assumes that `args` is non-empty.

Reply:
Whoops, thanks. Fixed!